src/job_token.rs

```diff
@@ -1,139 +1,163 @@
+        global_implicit_token: AtomicBool,
+        inner: sys::JobServerClient,
     }
 
-    // ... but if that fails for whatever reason select something
-    // reasonable and crate a new jobserver. Use `NUM_JOBS` if set (it's
-    // configured by Cargo) and otherwise just fall back to a
-    // semi-reasonable number. Note that we could use `num_cpus` here
-    // but it's an extra dependency that will almost never be used, so
-    // it's generally not too worth it.
-    let mut parallelism = 4;
-    if let Ok(amt) = env::var("NUM_JOBS") {
-        if let Ok(amt) = amt.parse() {
-            parallelism = amt;
+    impl JobServer {
+        pub(super) unsafe fn from_env() -> Option<Self> {
+            let var = var_os("CARGO_MAKEFLAGS")
+                .or_else(|| var_os("MAKEFLAGS"))
+                .or_else(|| var_os("MFLAGS"))?;
+
+            let inner = sys::JobServerClient::open(var)?;
+
+            Some(Self {
+                inner,
+                global_implicit_token: AtomicBool::new(true),
+            })
+        }
+
+        pub(super) fn try_acquire(&self) -> Result<Option<JobToken>, Error> {
+            if !self.global_implicit_token.swap(false, Relaxed) {
+                // Cold path, no global implicit token, obtain one
+                if self.inner.try_acquire()?.is_none() {
+                    return Ok(None);
+                }
+            }
+            Ok(Some(JobToken()))
+        }
+
+        pub(super) fn release_token_raw(&self) {
+            // All tokens will be put back into the jobserver immediately
+            // and they cannot be cached, since Rust does not call `Drop::drop`
+            // on global variables.
+            if self
+                .global_implicit_token
+                .compare_exchange(false, true, Relaxed, Relaxed)
+                .is_err()
+            {
+                // There's already a global implicit token, so this token must
+                // be released back into jobserver
+                let _ = self.inner.release();
+            }
         }
     }
+}
+
+mod inprocess_jobserver {
+    use super::JobToken;
+
+    use std::{
+        env::var,
+        sync::atomic::{AtomicU32, Ordering::Relaxed},
+    };
+
+    pub(crate) struct JobServer(AtomicU32);
 
-    // If we create our own jobserver then be sure to reserve one token
-    // for ourselves.
-    let client = jobserver::Client::new(parallelism).expect("failed to create jobserver");
-    client.acquire_raw().expect("failed to acquire initial");
-    return client;
+    impl JobServer {
+        pub(super) fn new() -> Self {
+            // Use `NUM_JOBS` if set (it's configured by Cargo) and otherwise
+            // just fall back to a semi-reasonable number.
+            //
+            // Note that we could use `num_cpus` here but it's an extra
+            // dependency that will almost never be used, so
+            // it's generally not too worth it.
+            let mut parallelism = 4;
```

Comment:
Yes, @dot-asm is against bumping msrv past 1.56

Reply:
@dot-asm can you make a defense as to why? It would be pretty nice to use available_parallelism here to avoid oversaturating the CPU on machines without a lot of concurrency (which includes a lot of vms/containers and machines people use for CI)

Reply:
Isn't `NUM_JOBS` going to be set most of the time when `cc` is being invoked through Cargo? Thus making `parallelism`'s default value less likely to be used.

Reply:
Yes, I've added a "TODO" comment.